kpi/models/asset.py

```diff
@@ -912,17 +980,66 @@ def to_clone_dict(self, version_uid=None, version=None):
     def to_ss_structure(self):
         return flatten_content(self.content, in_place=False)
 
+    def update_languages(self, children=None):
+        """
+        Updates object's languages by aggregating all its children's languages
+
+        Args:
+            children (list<Asset>): Optional. When specified, `children`'s languages
+            are merged with `self`'s languages. Otherwise, when it's `None`,
+            DB is fetched to build the list according to `self.children`
+
+        """
+        # If object is not a collection, it should not have any children.
+        # No need to go further.
+        if self.asset_type != ASSET_TYPE_COLLECTION:
+            return
+
+        obj_languages = self.summary.get('languages', [])
+        languages = set()
+
+        if children:
+            summaries = [child.summary for child in children]
+            languages = set(obj_languages)
+        else:
+            # ToDo validate whether all children should be included
+            # (e.g. archives?)
+            summaries = self.children.values_list('summary', flat=True)
+
+        for summary in summaries:
+            if isinstance(summary, str):
```

Comment:
we could add a TODO to simplify this once #2635 is done (or add a note in #2635 to remind us to come back here)

Reply:
(is `summary__languages` allowed with JSONB?)

Reply:
@jnm, I've tried to modified the code with `summary__languages` but it seems to be a bit slower than the current code.

```
        if children:
            languages = set(obj_languages)
            children_languages = [child.summary.get('languages')
                                  for child in children
                                  if child.summary.get('languages')]
        else:
            children_languages = list(self.children
                                      .values_list('summary__languages',
                                                   flat=True)
                                      .exclude(Q(summary__languages=[]) |
                                               Q(summary__languages=[None]))
                                      .order_by())

        # Flatten `children_languages` to 1-dimension list.
        if children_languages:
            languages.update(reduce(add, children_languages))

        languages.discard(None)
```

Fetching a nested property of a `JSONBField` has a cost in performance. 
```
In [15]: Asset.objects.filter(asset_type='collection').values_list('summary', flat=True).all()
Out[15]: SELECT "kpi_asset"."summary"
  FROM "kpi_asset"
 WHERE "kpi_asset"."asset_type" = 'collection'

Execution time: 0.001931s [Database: default]
```

```
In [14]: Asset.objects.filter(asset_type='collection').exclude(Q(summary__languages=[]) | Q(summary__languages=[None])).values_list('summary', flat=True).all()
Out[14]: SELECT "kpi_asset"."summary"
  FROM "kpi_asset"
 WHERE ("kpi_asset"."asset_type" = 'collection' AND NOT ((("kpi_asset"."summary" -> 'languages') = '[]' OR ("kpi_asset"."summary" -> 'languages') = '[null]')))

Execution time: 0.002602s [Database: default]
```

Reply:
I've run `./manage.py update_collection_languages` on my local env. 
See https://www.flowdock.com/rest/files/58517/chY2WGh8Vc8AB4YvMLe4Ew/update_languages.txt. 

I'll update the code in favor on this new version. It's clearer (IMHO) and the difference is insignificant.

Reply:
Cool :sunglasses: thanks for the extensive analysis :muscle: